outbox: Implement outbox for offline messages.

userAdityaa opened the pull request:
### How Does the Outbox Work?

* When the user is offline and attempts to send a message, instead of just displaying it in the chat with `resend` and `delete` options, the message is now also stored in the Outbox. This reassures the user that their message will be preserved and sent once they’re back online.
* Once the user goes from offline to online mode, the **new implementation** immediately tries to send the message, regardless of when it was originally created.
* If the message still fails to send due to technical issues, it remains in the Outbox. The user can then manually retry sending it using the `resend` button or choose to remove it entirely using the `delete` option.

---

Fixes: #32999 

<details>
<summary>Both `Drafts` and `Outbox` tab with empty message.</summary>
<img width="850" alt="Screenshot 2025-04-14 at 10 07 16 AM" src="https://github.com/user-attachments/assets/7e986307-85f4-4a1e-8bba-df201eed8c6c" />
<img width="866" alt="Screenshot 2025-04-14 at 10 03 23 AM" src="https://github.com/user-attachments/assets/0f474d85-5319-41bb-bc42-d7c115d78382" />
</details>





<details>
<summary>`Drafts` tab with message and `Outbox` tab with no message.</summary>
<img width="850" alt="Screenshot 2025-04-14 at 10 07 59 AM" src="https://github.com/user-attachments/assets/4896cc40-b153-4d7e-8752-2dae79580e92" />
<img width="850" alt="Screenshot 2025-04-14 at 10 08 10 AM" src="https://github.com/user-attachments/assets/6f15e4c5-831a-4bdd-bae8-8d58e9b5edfb" />
</details>

<details>
<summary>Both `Drafts` and `Outbox` tab with some message.</summary>
<img width="850" alt="Screenshot 2025-04-14 at 10 07 59 AM" src="https://github.com/user-attachments/assets/4896cc40-b153-4d7e-8752-2dae79580e92" />
<img width="850" alt="Screenshot 2025-04-14 at 10 10 54 AM" src="https://github.com/user-attachments/assets/186d80a0-4a4a-4b22-ba6e-5556c67597f9" />
</details>

<details>
<summary>Tooltip for saving as Drafts and Outbox.</summary>
<img width="419" alt="Screenshot 2025-04-14 at 10 36 17 AM" src="https://github.com/user-attachments/assets/199ac582-c167-4ffd-b14f-b211b8e3ee5f" />
<img width="406" alt="Screenshot 2025-04-14 at 12 26 48 PM" src="https://github.com/user-attachments/assets/12675d79-e0fc-4497-ae28-a7350cbb2a05" />
</details>

<details>
<summary>The function attempts to send the message once the app is back online.</summary>

https://github.com/user-attachments/assets/84a93417-20a0-4853-8982-c50cf87b1dc5


</details>


<details>
<summary>The message is stored in the outbox when the app is reloaded.</summary>

https://github.com/user-attachments/assets/7bed4ebe-9d87-494b-96ef-b9002f1eabd6


</details>


<details>
<summary>When the user attempts to delete or resend the message from the outbox overlay after coming back online.</summary>

https://github.com/user-attachments/assets/42448aa3-65cb-491d-937e-8d580ff4b592


https://github.com/user-attachments/assets/d2e3849c-0131-4046-9b4f-f056d4730729


</details>

<details>
<summary>Deleting the message through either the outbox overlay or directly from the message feed.</summary>

https://github.com/user-attachments/assets/c1e4621a-5e4e-45d8-a1e1-0410895e5628

https://github.com/user-attachments/assets/e8e77982-99dd-478c-ac35-86ade2aae73c




</details>


<details>
<summary>Bulk deletion of the outbox, both while offline and online. </summary>

https://github.com/user-attachments/assets/0be60508-d9c3-47d2-bb51-e3470a8ebc2a


https://github.com/user-attachments/assets/64ed0e01-eb1a-414c-b326-220191544396



</details>


<details>
<summary>Implementation to handle the race condition that occurs when the user manually clicks the resend button while the now Implemented resend function is already in the process of sending the message.</summary>

https://github.com/user-attachments/assets/be77f39f-ec92-49dc-b1fd-20ecc9ed66eb


</details>



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

userAdityaa committed:
outbox: Implement outbox for offline messages.

This commit adds a new feature called `Outbox` which
replicates drafts, but for offline messages. Outbox saves
the offline un-send message and ensures user about not
losing the message. The user can either `resend` the
message or `delete` the message as per the UI in the
outbox tab.

Fixes #32999.

userAdityaa commented:
Currently there are conflicting changes going on with respect to this PR, as we are updating the drafts tooltip and also we are updating the popover header and footer UI as the issue is already being tracked. I have opened this for the feedback on the logic implemented.

alya commented:
I don't understand the comment above, sorry. What are you saying about conflicting changes?

userAdityaa commented:
I should have explained it better earlier. We’re currently tracking multiple UI changes related to a redesign in [this issue](https://github.com/zulip/zulip/issues/34377). Since this PR also includes UI updates connected to that, it will cause conflicts.

userAdityaa commented:
While doing a final check, there still one issue that is not solved yet, so I'm converting this to draft. Currently, I have my exams going on so. I will mark this for review in few days as soon as I can.

userAdityaa commented:
This is now ready for review and feedback. I've implemented everything I could think of related to the outbox, and I've also updated the PR description with videos covering all the cases. 

If there is any issue, I will look into it and solve it accordingly.

amanagr commented:
Closing in favour of #38356
